### PR TITLE
Implement ability to list addresses in the wallet layer, API and CLI

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#220 

# Overview

- [ ] I have implemented the ability to list addresses in the wallet layer, API and CLI

# Comments

This would make it easier for @piotr-iohk and us to write some integration tests for the transaction API. We could have derived addresses by hand, but this achieves to complete the high-priority endpoints of the API.

NOTE: It's a "partial" implementation which for now, ignores the filtering based on the address state (`?state=used` or `?state=unused`). This can be implemented later.

:warning: I recommend looking at the change commit by commit ;)

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery.hs [hunk 1]
```diff
@@ -156,6 +158,39 @@ class GenChange s where
...
+        :: s
+        -> Address
+        -> Address
+        -> Ordering
+
+-- | Extract the list of all known addresses.
+--
+-- NOTE: Change addresses aren't considered "known" until they've been used. The
+-- rationale is that, we don't want users or consumers of the wallet to be using
+-- change addresses primaturely.
```
Comment (akegalj) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery.hs]: s/primaturely/prematurely

File: lib/http-bridge/test/integration/Test/Integration/Scenario/CLISpec.hs [hunk 2]
```diff
@@ -119,6 +119,14 @@ specWithCluster = do
         err `shouldBe` "Ok.\n"
         out `shouldNotContain` "CLI Wallet"
         c `shouldBe` ExitSuccess
+
+    it "CLI - Can list addresses" $ \ctx -> do
+        walId <- createWallet ctx "CLI Wallet" mnemonics15
```
Comment (akegalj) [inline lib/http-bridge/test/integration/Test/Integration/Scenario/CLISpec.hs:124]: why don't we also generate `mnemonics` here?

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs [hunk 3]
```diff
@@ -312,6 +332,77 @@ prop_lookupDiscovered (s0, addr) =
...
+        (False,_) -> cover 5 True "known-unknown" $ prop LT
+        (True, _) -> property True
+  where
+    prop ordering = compareDiscovery s known addr === ordering
+
+prop_compareAntiSymmetric
+    :: (SeqState DummyTarget, ShowFmt Address, ShowFmt Address)
+    -> Property
+prop_compareAntiSymmetric (s, ShowFmt a1, ShowFmt a2) =
+    fst (isOurs a1 s) && fst (isOurs a2 s) ==> cover 90 (a1 /= a2) "a1 /= a2" prop
```
Comment (akegalj) [inline lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs]: I wonder what is distribution to pick two `isOurs` wallets (hopefully precondition won't take too long - depending on the generator)

**Review by akegalj — APPROVED** [review]
lgtm

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs (hunk 3 above)
Comment (akegalj) [reply lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs]:
```
      1) Cardano.Wallet.Primitive.AddressDiscovery.CompareDiscovery compareDiscovery is anti-symmetric
           *** Gave up! Passed only 199 tests; 232 discarded tests (98.5% a1 /= a2).
```
almost :P

File: lib/http-bridge/test/integration/Test/Integration/Scenario/CLISpec.hs (hunk 2 above)
Comment (KtorZ) [reply lib/http-bridge/test/integration/Test/Integration/Scenario/CLISpec.hs:124]: Hmmm. Good question. No particular reasons that I can think of.

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs (hunk 3 above)
Comment (KtorZ) [reply lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs]: Ho damn :o ! The coverage is too tight maybe here ^^"

File: lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs [hunk 4]
```diff
@@ -312,6 +321,32 @@ prop_lookupDiscovered (s0, addr) =
...
+{-------------------------------------------------------------------------------
+                        Properties for CompareDiscovery
+-------------------------------------------------------------------------------}
+
+prop_compareKnownUnknown
+    :: (SeqState DummyTarget, ShowFmt Address, ShowFmt Address)
+    -> Property
+prop_compareKnownUnknown (s, ShowFmt known, ShowFmt addr) =
+    case (fst $ isOurs known s, fst $ isOurs addr s) of
+        (True, False) -> cover 10 True "known-unknown" $ prop LT
```
Comment (paweljakubas) [inline lib/core/test/unit/Cardano/Wallet/Primitive/AddressDiscoverySpec.hs:344]: this one along with `prop_compareAntiSymmetric` will check `known-unknow` and `unknow-know`. `know-know` will be covered by the aforementioned property. The missing one is `unknow-unknow` although very obvious one

**Review by paweljakubas — APPROVED** [review]
LGTM!

**Pull request merged**